src/hypofuzz/provider.py

```diff
@@ -26,29 +80,409 @@ def fresh_choice(
+
+        return Report(
+            database_key=b64encode(self.database_key).decode(),
+            nodeid=self.nodeid,
+            elapsed_time=self.elapsed_time,
+            timestamp=time.time(),
+            worker_uuid=self.worker_identity.uuid,
+            status_counts=StatusCounts(self.status_counts),
+            behaviors=len(self.corpus.behavior_counts),
+            fingerprints=len(self.corpus.fingerprints),
+            since_new_branch=(
+                None
+                if (self.ninputs == 0 or self.corpus.interesting_examples)
+                else self.since_new_branch
+            ),
+            phase=self.phase,
+        )
+
+    @contextmanager
+    def per_test_case_context_manager(self) -> Generator[None, None, None]:
+        # note that this only collects coverage during test execution, not
+        # prep_args_kwargs_from_strategies. If we want to collect coverage during
+        # input generation, we'll need a separate hypothesis-side hook.
+        self.before_test_case()
+        try:
+            with self.collector or CoverageCollector() as collector:
+                yield
+        finally:
+            assert self._state is not None
+            self._state.branches = frozenset(collector.branches)
+            # after_test_case is called by on_observation, since we want to wait
+            # until we know a few things about how the execution went
+
+    def before_test_case(self) -> None:
+        if not self._started:
+            self._startup()
+
+        assert self.corpus is not None
+        replay_priority = None
+        extra_queue_data = None
+        while self._replay_queue:
+            (replay_priority, choices, extra_queue_data) = self._replay_queue.pop()
+
+            # we checked if we had this choice sequence when we put it into the
+            # replay_queue on on_event, but we check again here, since we might have
+            # executed it in the interim.
+            if (
+                # Choices doesn't handle ChoiceTemplate
+                all(not isinstance(choice, ChoiceTemplate) for choice in choices)
+                and Choices(choices) in self.corpus.corpus
+            ):
+                continue
+
+            self._start_phase(Phase.REPLAY)
+            break
+        else:
+            # We stay in blackbox mode until we've generated 1000 consecutive examples
+            # without new coverage, and then switch to mutation.
+
+            # TODO: currently hard-coding a particular mutator; we want to do MOpt-style
+            # adaptive weighting of all the different mutators we could use.
+            # For now though, we'll just use a hardcoded swapover point
+            if self.since_new_branch < 1000 or self.random.random() < 0.05:
+                mutator: Mutator = BlackBoxMutator(self.corpus, self.random)
+            else:
+                mutator = CrossOverMutator(self.corpus, self.random)
```

Comment:
Let's use a less dumb / more-smoothly-adaptive option?  

e.g. "use crossover with probability `(ninputs - len(corpus)) / ninputs`"

eventually we'll definitely want MOpt-style estimators, but this is still a huge improvement for now